### PR TITLE
Express: improve `req.query` type

### PR DESCRIPTION
**Pull request by OliverJAsh**
- `any` disables all type checking. `unknown` should be used instead:
  ```ts
  declare const any: any;
  any.foo.bar; // no compile error, but runtime exception!
  ```
- We can't use `unknown` because these types don't yet target TS 3+, so I used `PoorMansUnknown` (`{} | null | undefined`) instead, which achieves the same thing.
- [The docs suggest `req.query` will always be an object](https://expressjs.com/en/api.html#req.query).
- The exact type of the object values is unknown, because it can be configured via [`query parser`](https://expressjs.com/en/api.html#app-settings-property).

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
- - https://expressjs.com/en/api.html#req.query
- - https://expressjs.com/en/api.html#app-settings-property (see `query parser`)
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by puneetar** [comment]
@OliverJAsh  This is definitely an improvement over `any`. 
It would be great if we can make have the query as generic similar to `params: P`.

I will give it a stab later today.

**Comment by puneetar** [comment]
Wait, @OliverJAsh  Is this a breaking change? Will it break types for the projects using this library?

**Comment by puneetar** [comment]
I created a PR where we can have the query param as generic and let the consumer specify the type. 
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/43434

**Comment by OliverJAsh** [comment]
> Wait, @OliverJAsh Is this a breaking change? Will it break types for the projects using this library?

Any type change is a breaking change. In this case, it should only break projects which have bugs, I think?

**Comment by puneetar** [comment]
> In this case, it should only break projects which have bugs, I think?

No, The compilers will start throwing errors as we are updating `any` to `poorMansUndefined`.   In case of `any` we are making assumptions on the query type, those assumptions may still be correct but after this update they need to explicitly use type refinement to fix the compile error. 

IMO, using `poorMansUndefined` forces one to use type refinement, but using generics will allow them to make safe assumptions about the query parameters.

**Comment by OliverJAsh** [comment]
I see your point re. needing to add refinements!

I added the generics for params, and in hindsight I'm not sure it was worth it. The generic is really awkward to use because you have to provide other all generics at the same time. A type assertion would work just as well IMO:

```typescript
type MyQuery = {
    q: string;
};

app.get('/:foo', req => {
    const query = req.query as MyQuery

    query.q; // $ExpectType string
    query.a; // $ExpectError
})

// vs

// Query can be a custom type
app.get<{}, any, any, MyQuery>('/:foo', req => {
    req.query.q; // $ExpectType string
    req.query.a; // $ExpectError
});
```

For that reason I would personally prefer we avoid adding another generic and just encourage type assertions instead.

File: types/express-serve-static-core/index.d.ts [hunk 1]
```diff
@@ -463,7 +463,12 @@ export interface Request<P extends Params = ParamsDictionary, ResBody = any, Req
 
     params: P;
 
-    query: any;
+    query: {
+        [key: string]:
+            // The value type here is a "poor man's `unknown`". When these types support TypeScript
+            // 3.0+, we can replace this with `unknown`.
+            {} | null | undefined
+    };
```
Comment (OliverJAsh) [inline types/express-serve-static-core/index.d.ts]: Alternatively we could use this type:

```ts
type ParsedQs = { [key: string]: string | string[] | ParsedQs };
```

That's the return type of `qs.parse` which is the default query parser: https://expressjs.com/en/api.html#app.settings.table

If a custom query parser is used, users can add assert the type:

```typescript
type MyQuery = {
    q: number;
};

app.get('/', req => {
    const query = req.query as MyQuery

    query.q; // $ExpectType number
})
```

**Comment by puneetar** [comment]
Adding generics will allow both the examples you specified above. If one wants to use type assertion they could use that, but also pass in the proper generics. 

I am library that I use internally, we aim on using all the generics for a Request type, In that case this would work perfectly.

**Comment by uniqueiniquity** [comment]
@puneetar is this ready to be merged?

**Comment by puneetar** [comment]
@uniqueiniquity I don't think this should be merged in the current state. 

We have another PR in progress to fix this issue https://github.com/DefinitelyTyped/DefinitelyTyped/pull/43434 PR.

**Comment by uniqueiniquity** [comment]
@puneetar Alright, sounds good.
To keep things clean, I'm going to close this PR.
If it needs to be revived, please fork and reopen.

**Pull request closed without merging**